### PR TITLE
config: switch to a upgradeable parameter group format

### PR DESCRIPTION
**Pull request by nzmichaelh**
This is a proof of concept and a follow up to #963.  If it looks good then I'll split off the unrelated changes and finish the corner cases.

This allows upgrading and downgrading of the config without losing all settings.  Parameters are organised as parameter groups and stored separately.  New parameters can be added to the end of a parameter group and will be automatically upgraded.  New parameter groups can be added at any time.

The idea is to use parameter groups *everywhere* so the protocol format is the same as the storage format and the in-RAM format is the same as the protocol format.  In most cases this means we won't need encoders and decoders.

Parameter groups are defined as structs so (hopefully) there's no code generation for the flight software.  I also have a parser that parses the structs and embedded comments and can be used to automatically generate protocol wrappers for Baseflight.

The concept is based on SAE J1939.

So, epic time.  Thoughts?

### CODE REVIEW

**Comment by digitalentity** [comment]
If I understand correctly, if I add a new parameter to some PG, only that PG would be reset to defaults? That's great news!

**Comment by nzmichaelh** [comment]
@digitalentity  Ah, it's even better than that: only the new parameter gets defaulted.

**Comment by nzmichaelh** [comment]
(providing you add the parameter to the end of a group)

**Comment by digitalentity** [comment]
We must take into account that parameters can be also replaced by a new parameter of the same byte size but with a different meaning. In such case modified PG must also be reset to defaults to avoid parameter corruption.

**Comment by nzmichaelh** [comment]
It's generally better to deprecate the old field and add the replacement field at the end, especially if you want backwards protocol compatibility.

Say you want to have more range on the ki term on a PID controller.  The current layout is:

struct PIDConfig { uint8 kp, uint8 ki, uint8 kd }

I'd recommend changing this to:

struct PIDConfig { uint8 kp, uint8 deprecated_ki, uint8 kd, uint16 ki }

**Comment by digitalentity** [comment]
Good idea, but this would increase eeprom usage and we are already at the limit of 2K page.

**Comment by nzmichaelh** [comment]
Hmm, I wonder how you could do cheap group level format change detection?  The types of change are:

1. Append a field
2. Change the width of a field
3. Reorder fields
4. Insert or delete a field
5. Change the meaning of a field

(1) is handled by this patch.

(2) could be handled by, say, varint encoding but that means you need an encoder + decoder instead of a memcpy.  That might be fine, but you're trading more flash usage for less EEPROM-ish usage.

The rest are tricky.  One way is to store a format version with each group, but that increases the EEPROM usage.  There might be enough room in the current four byte header to slip that in.  We could automatically generate the version number by using the struct parser I talked about as it can hash the field names.

Another way is to tag the parameters with a field ID, but that means you need to store the tag.  You could also say that the memory address is the tag, but then you need a decoder to copy those back out to the right place in RAM.

Short story: we can handle all the modification cases, but it's more complex (which is fine), and could take more flash space overall.

**Comment by nzmichaelh** [comment]
Oh!  It's also possible to grow the size of the EEPROM-ish space without resetting to defaults.  The groups are small enough that you could fill the top 2 kB and then, in a later version, overflow into the second to top 2 kB.

**Comment by digitalentity** [comment]
(2), (3) and (4) can probably be handled by a simple 1-byte CRC of a group. PGN can probably be made an uint8_t this leaves us 1 byte for CRC.
(5) is tricky and should be left to developer, changing the meaning of the field.

**Comment by nzmichaelh** [comment]
One goal is to make the in-RAM, in-flash, and protocol format the same to keep the code simple and reduce the code size.  If there's some type of format field then this will also need to go out on the wire (which is fine).

The format number could be generated at compile time by hashing the field types and names.  Flatbuffers uses this method.  The format shouldn't change if you make a compatible change like appending a field, so you'd need to keep the old format in a database and only generate a different number if there was a structural change.  Assuming that format changes are rare, then it's probably better to have the developer update the format number manually instead.

Assuming format changes are rare, how about making the developer pick a new PGN instead?

**Comment by digitalentity** [comment]
I like the idead of making the developer pick a new PGN when significantly changing format. However as one of the developers I vote for an compile-time option controling the behavior of new code when loading PG with different size. With this option disabled only new parameters in PG will get defaulted, but with this option enabled the whole PG will be reset to defaults. 
You see, I sometimes add and remove parameters not only at the end of a conf struct (a PG) and I would like to avoid resetting the whole config and be able to reset only one PG.

**Comment by nzmichaelh** [comment]
Sounds good.  I'm not sure how that will map through to the protocol: if you change the format, then either the receiver needs to know the format in addition to the PGN, or the board needs to encode to the old format before sending.

I can add a format field per PGN (it's good practice in any case) and we can worry about the protocol problem if and when it occurs.

/me gets to it.

**Comment by digitalentity** [comment]
You'll probably want to have look at #968. A parameter ID can be somehow related to PGN

**Comment by nzmichaelh** [comment]
@ledvinap yip, it's flash.  The code calls the in-flash configuration storage 'EEPROM' due to its AVR heritage.

**Comment by nzmichaelh** [comment]
@digitalentity yip, J1939 also assigns each individual parameter a 'suspect parameter number' (SPN).  It's called that as it was originally used for the engine fault codes, as in 'engine temperature, which is field 5 of PGN 123 is suspect.'

**Comment by nzmichaelh** [comment]
Update: I've added tests for the save / load logic.  Next step is to start breaking up masterConfig.

**Comment by nzmichaelh** [comment]
@digitalentity it's looking pretty complete.  Have a look at 1b06346 for what moving a config item to a parameter group involves.

If this looks good then I'll start tidying things up and migrating more of the config to parameter groups.

**Comment by digitalentity** [comment]
@nzmichaelh 
That's great news! As I can see we also get rid of lots of struct pointer dereference operations and greatly simplify MSP handing. Such approach can actually save us a few cpu cycles and some flash memory. Keep up the good work! Hopefully @hydra approves this

**Comment by digitalentity** [comment]
@nzmichaelh 
How is this going?

**Comment by nzmichaelh** [comment]
Hi @digitalentity, thanks for the ping - I'm happy with the basic design and implementation.  I paused as there's a fair amount of rework on the rest of Cleanflight past this to make use of it.  It's lots of small things like:

* Splitting parameters out of masterConfig and into the module that uses it
* Checking that every parameter struct has the fields in the right order (as the memory format must match the protocol format)
* Splitting some big parameter structs into smaller, related structs (again as the memory format must match the protocol format)

@hydra would you mind having a look at the design and implementation and seeing what you think?  If it looks good then I'll split this patch into a few, incremental patches.  We can then make masterConfig a parameter group, add new modules as their own parameter groups, and (slowly?) refactor masterConfig.

**Comment by hydra** [comment]
@nzmichaelh I had a look at this and I must say I prefer it over the giant configuration structure that was inherited from baseflight.  being able to arbitrarily store anything we like is great.

With regards to 2k limits, the most targets now use 4KB. so that issue is moot.

There is currently some work into extending the GUI to make other configuration changes and I see this change as being useful there.

Any chance you could do a rebase against master?

**Comment by hydra** [comment]
@digitalentity @nzmichaelh p.s.  sorry for not commenting on this earlier.

**Comment by hydra** [comment]
I was thinking about this a bit more today, I think that if we took this approach then the main `master_t` and `profile_t` would just disappear and suddenly we would be able to decouple even more things and reduce the header file inclusions in some areas.

**Comment by hydra** [comment]
I did a bunch of work yesterday on a new project that uses a bunch of code from the cleanflight codebase and wish i had this already.

@nzmichaelh any chance you could rebase this against latest master?

**Comment by nzmichaelh** [comment]
@hydra sorry for the delay in responding.  I've switched over to to ArduRover / Pixhawk for my project.  You're welcome to grab this pull request and finish it off though!

**Comment by hydra** [comment]
@nzmichaelh ok, thanks for the update.  gonna look at this again now.

**Comment by hydra** [comment]
ok, i have this working against current master now.  I'll create a new branch with the changes in it once it's ready.  obviously only the failsafe and boardAlignment groups are done now, the rest will follow in additional commits to the new branch.

File: src/main/io/serial_msp.c [hunk 1]
```diff
@@ -695,6 +522,13 @@ static bool processOutCommand(uint8_t cmdMSP)
     uint8_t wp_no;
     int32_t lat = 0, lon = 0;
 #endif
+    const pgRegistry_t *reg = pgFind(cmdMSP);
+
+    if (reg != NULL) {
+        headSerialReply(reg->size);
```
Comment (hydra) [inline src/main/io/serial_msp.c]: this line was not needed. s_struct() sends the size.

File: src/main/io/serial_msp.c [hunk 2]
```diff
@@ -1248,6 +1068,13 @@ static bool processInCommand(void)
     int32_t lat = 0, lon = 0, alt = 0;
 #endif
 
+    const pgRegistry_t *reg = pgFindForSet(currentPort->cmdMSP);
+
+    if (reg != NULL) {
+        pgLoad(reg, currentPort->inBuf + currentPort->indRX, currentPort->dataSize);
+        return true;
```
Comment (hydra) [inline src/main/io/serial_msp.c]: this was missing `headSerialReply(0);` before `return true;`

File: src/main/config/config_streamer.c [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    const uint8_t *pat = p;
+    const uint8_t *pend = pat + size;
+
+    for (; pat != pend; pat++) {
+        c->buffer.b[c->at++] = *pat;
+        c->chk ^= *pat;
+
+        if (c->at == sizeof(c->buffer)) {
+            c->err = write_word(c, c->buffer.w);
+            c->at = 0;
```
Comment (hydra) [inline src/main/config/config_streamer.c]: `c->err` can be overwritten if size is large enough.  not 100% certain about the implementation of `write_word()` and the the handling of `c->err`.

**Comment by hydra** [comment]
created branch https://github.com/cleanflight/cleanflight/commits/configuration-storage

**Comment by hydra** [comment]
discussion and additional commits should take place on the new PR #1976

@nzmichaelh thanks for all your hard work on this.  if you would like to help out further please checkout the new branch and create pull requests that target this new branch so they can be merged into PR #1976

Cheers!

**Pull request closed without merging**